### PR TITLE
[C/PyTorch] RoPE fixes and minor improvements for fused attention

### PR DESCRIPTION
**Pull request by cyanguwa**
This PR 
- adjusts the support for fused attention in preparation of upcoming cuDNN releases,
- reduces the number of unit tests for fused attention, and
- fixes rotary position embedding issue #455 .

### CODE REVIEW

**Comment by cyanguwa** [comment]
Pipeline 10079370

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -691,6 +692,64 @@ def flash_attn_forward_func_with_cp(q, k, v, cu_seqlens_q, cu_seqlens_k,
...
 
+class RotaryPositionEmbedding(torch.nn.Module):
+    """
+    Implements Rotary Position Embedding from https://arxiv.org/abs/2104.09864.
+    """
+    def __init__(
+        self,
+        dim: int,
+        seq_len_interpolation_factor: int = None,
+        pretrained_max_position_embeddings: int = None,
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:703]:
```suggestion
        seq_len_interpolation_factor: Optional[int] = None,
        pretrained_max_position_embeddings: Optional[int] = None,
```

File: transformer_engine/pytorch/attention.py [hunk 2]
```diff
@@ -691,6 +692,64 @@ def flash_attn_forward_func_with_cp(q, k, v, cu_seqlens_q, cu_seqlens_k,
...
+            else:
+                # fixed linear scaling
+                seq *= 1 / self.seq_len_interpolation_factor
+
+        freqs = torch.einsum('i , j -> i j', seq, self.inv_freq)
+        # first part even vector components, second part odd vector components,
+        #  2 * dim in dimension size
+        emb = torch.cat((freqs, freqs), dim=-1)
+        # emb [seq_length, .., dim]
+        return rearrange(emb, 'n d -> n 1 1 d')
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:751]: Could we remove the einops dependency?
```suggestion
        return emb.reshape(emb.size(0), 1, 1, emb.size(1))
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/attention.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 from typing import Any, Callable, List, Optional, Tuple, Union, Dict
 from pkg_resources import packaging
 import numpy as np
+from einops import rearrange
```
[the remark this answers is not in the document]
Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:14]: Let's use `.transpose` instead @cyanguwa

File: transformer_engine/pytorch/attention.py (hunk 1 above)
Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py:703]: Done

File: transformer_engine/pytorch/attention.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py]: Thanks - I'm adopting Tim's suggestion.

File: transformer_engine/pytorch/attention.py (hunk 2 above)
Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py]: Thanks!

**Review by timmoon10 — APPROVED** [review]
LGTM

**Review by ksivaman — APPROVED** [review]
LGTM

**Comment by cyanguwa** [comment]
All CI has passed (10173716), merging.

**Pull request merged**